server/etcdserver/apply/apply_kv_test.go

```diff
@@ -0,0 +1,414 @@
+		Key: []byte(key4),
+	})
+	require.NoError(t, err)
+	require.EqualValues(t, 1, respDelete.Deleted)
+
+	// deleting non-existing key generates no error, just `Deleted: 0` response
+	respDelete, err = appV3be.DeleteRange(nil, &pb.DeleteRangeRequest{
+		Key: []byte(key),
+	})
+	require.NoError(t, err)
+	require.EqualValues(t, respDelete.Deleted, 0)
+}
+
+func TestRangeOptions(t *testing.T) {
+	appV3be, _ := defaultApplierV3Backend(t)
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	putKeys(ctx, t, fourKeys, appV3be) // revs 2, 3, 4 and 5
+	putKeys(ctx, t, []pb.PutRequest{
+		{Key: []byte(key), Value: updatedValue},  // rev 6
+		{Key: []byte(key), Value: updatedValue2}, // rev 7
+	}, appV3be)
+
+	tcs := []struct {
+		name       string
+		request    *pb.RangeRequest
+		expectedV  [][]byte
+		expectedC  int64
+		customTest func(r *pb.RangeResponse)
+	}{
+		{
+			name:      "Selecting single key revision",
+			request:   &pb.RangeRequest{Key: []byte(key), Revision: 6},
+			expectedV: [][]byte{updatedValue},
+			expectedC: 1,
+		},
+		{
+			// Should yield 3 KVs and count:4, since `key` has a revision6 value, but it's current ModRevision is 7
+			name:      "All keys in ModRevision range",
+			request:   &pb.RangeRequest{Key: zero, RangeEnd: zero, MinModRevision: 3, MaxModRevision: 6},
+			expectedV: [][]byte{fourKeys[1].Value, fourKeys[2].Value, fourKeys[3].Value},
+			expectedC: 4,
+		},
+		{
+			name:      "All keys in CreateRevision range",
+			request:   &pb.RangeRequest{Key: zero, RangeEnd: zero, MinCreateRevision: 2, MaxCreateRevision: 5},
+			expectedV: [][]byte{updatedValue2, fourKeys[1].Value, fourKeys[2].Value, fourKeys[3].Value},
+			expectedC: 4,
+		},
+		{
+			name:       "Limited subset of all keys sorted desc by value",
+			request:    &pb.RangeRequest{Key: zero, RangeEnd: zero, Limit: 3, SortOrder: pb.RangeRequest_DESCEND, SortTarget: pb.RangeRequest_VALUE},
+			expectedV:  [][]byte{updatedValue2, fourKeys[3].Value, fourKeys[2].Value},
+			expectedC:  4,
+			customTest: func(r *pb.RangeResponse) { require.True(t, r.More) },
+		},
+		{
+			name:      "Limited subset of keys sorted asc by version",
+			request:   &pb.RangeRequest{Key: zero, RangeEnd: fourKeys[2].Key, SortOrder: pb.RangeRequest_ASCEND, SortTarget: pb.RangeRequest_VERSION},
+			expectedV: [][]byte{fourKeys[1].Value, updatedValue2},
+			expectedC: 2,
+		},
+		{
+			name:       "All keys only",
+			request:    &pb.RangeRequest{Key: zero, RangeEnd: zero, KeysOnly: true},
+			expectedV:  [][]byte{nil, nil, nil, nil},
+			expectedC:  4,
+			customTest: func(r *pb.RangeResponse) { require.Len(t, r.Kvs, 4) },
+		},
+		{
+			name:      "Count keys greater then or equal to key2",
+			request:   &pb.RangeRequest{Key: []byte(key2), RangeEnd: zero, CountOnly: true},
+			expectedV: [][]byte{},
+			expectedC: 3,
+		},
+		// TODO: test Serializable
+	}
+
+	for _, tc := range tcs {
```

Comment:
please use `tc := tc` before `t.Run`, just in case that the previous var has been shadowed by next.

Reply:
First of all, thank you for the nice catch!
Can you please increment the discussion further just for me to get it right?

🤯 ... As the `for` block declares an anonymous function capturing `tc` can't we assume the `for` bock captures `tc`? I mean, doesn't the `for` block hold it's own reference to `tc` value  (from the moment of the blocks creation)?

Are you using any static analysis to detect those potential flaws?

Reply:
Becasue the address of `tc` is the same to each loop. If there are two goroutines to read the same var, it's not expected. Based on your current code, the `t.Run` is running in sync. It can work well. Using `tc := tc` is to prevent from being shadowed if we enable `t.Parallel` in the sub-test.

> Are you using any static analysis to detect those potential flaws?

As far as I know, I don't see static analysis about this. :(

> NOTE: You can checkout this link https://github.com/golang/go/wiki/LoopvarExperiment. Hope it can help.